crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -0,0 +1,1285 @@
+    ///
+    /// Args:
+    ///     instruction (Union[qiskit.circuit.Instruction, Type[qiskit.circuit.Instruction]]):
+    ///         The operation object to add to the map. If it's parameterized any value
+    ///         of the parameter can be set. Optionally for variable width
+    ///         instructions (such as control flow operations such as :class:`~.ForLoop` or
+    ///         :class:`~MCXGate`) you can specify the class. If the class is specified than the
+    ///         ``name`` argument must be specified. When a class is used the gate is treated as global
+    ///         and not having any properties set.
+    ///     properties (dict): A dictionary of qarg entries to an
+    ///         :class:`~qiskit.transpiler.InstructionProperties` object for that
+    ///         instruction implementation on the backend. Properties are optional
+    ///         for any instruction implementation, if there are no
+    ///         :class:`~qiskit.transpiler.InstructionProperties` available for the
+    ///         backend the value can be None. If there are no constraints on the
+    ///         instruction (as in a noiseless/ideal simulation) this can be set to
+    ///         ``{None, None}`` which will indicate it runs on all qubits (or all
+    ///         available permutations of qubits for multi-qubit gates). The first
+    ///         ``None`` indicates it applies to all qubits and the second ``None``
+    ///         indicates there are no
+    ///         :class:`~qiskit.transpiler.InstructionProperties` for the
+    ///         instruction. By default, if properties is not set it is equivalent to
+    ///         passing ``{None: None}``.
+    ///     name (str): An optional name to use for identifying the instruction. If not
+    ///         specified the :attr:`~qiskit.circuit.Instruction.name` attribute
+    ///         of ``gate`` will be used. All gates in the ``Target`` need unique
+    ///         names. Backends can differentiate between different
+    ///         parameterization of a single gate by providing a unique name for
+    ///         each (e.g. `"rx30"`, `"rx60", ``"rx90"`` similar to the example in the
+    ///         documentation for the :class:`~qiskit.transpiler.Target` class).
+    /// Raises:
+    ///     AttributeError: If gate is already in map
+    ///     TranspilerError: If an operation class is passed in for ``instruction`` and no name
+    ///         is specified or ``properties`` is set.
+    #[pyo3(signature = (instruction, name, properties=None))]
+    fn add_instruction(
+        &mut self,
+        instruction: TargetOperation,
+        name: String,
+        properties: Option<PropsMap>,
+    ) -> PyResult<()> {
+        if self.gate_map.contains_key(&name) {
+            return Err(PyAttributeError::new_err(format!(
+                "Instruction {:?} is already in the target",
+                name
+            )));
+        }
+        let mut qargs_val: PropsMap = PropsMap::default();
```

Comment:
Could we use `with_capacity()` here since  this will be unused or have `properties.len()`?